File: CS/ReportAtRuntimeMvcApp/Services/CustomReportProvider.cs
using DevExpress.DataAccess.Sql;
using DevExpress.XtraPrinting;
using DevExpress.XtraReports.Services;
using DevExpress.XtraReports.UI;
using System.Drawing;

namespace ReportAtRuntimeMvcApp.Services
{
    public class CustomReportProvider : DevExpress.XtraReports.Services.IReportProvider
    {
        public XtraReport GetReport(string id, ReportProviderContext context)
        {
            if (string.IsNullOrEmpty(id))
                return null;
            switch (id)
            {
                case "TestReport":
                    return CreateProductsReport();
            }
            return new XtraReport();
        }
        XtraReport CreateProductsReport()
        {
            XtraReport report = new XtraReport();
            ReportHeaderBand headerBand = new ReportHeaderBand()
            {
                HeightF = 80
            };
            report.Bands.Add(headerBand);
            headerBand.Controls.Add(new XRLabel()
            {
                Text = "Categories Report",
                SizeF = new SizeF(650, 80),
                TextAlignment = TextAlignment.BottomCenter,
                Font = new Font("Arial", 36)
            });
            DetailBand detailBand = new DetailBand();
            report.Bands.Add(detailBand);
            XRPictureBox pbPicture = new XRPictureBox()
            {
                LocationF = new PointF(10, 10),
                SizeF = new SizeF(190, 90),
                Sizing = ImageSizeMode.ZoomImage
            };
            pbPicture.ExpressionBindings.Add(new ExpressionBinding("Image", "Picture"));
            detailBand.Controls.Add(pbPicture);
            XRLabel lbCategoryName = new XRLabel()
            {
                LocationF = new PointF(200, 10),
                SizeF = new SizeF(440, 50),
                TextAlignment = TextAlignment.BottomLeft,
                Font = new Font("Arial", 24)
            };
            lbCategoryName.ExpressionBindings.Add(new ExpressionBinding("Text", "CategoryName"));
            detailBand.Controls.Add(lbCategoryName);
            XRLabel lbDescription = new XRLabel()
            {
                LocationF = new PointF(200, 60),
                SizeF = new SizeF(440, 40),
                TextAlignment = TextAlignment.TopLeft,
                Font = new Font("Arial", 14, FontStyle.Italic)
            };
            lbDescription.ExpressionBindings.Add(new ExpressionBinding("Text", "Description"));
            detailBand.Controls.Add(lbDescription);
            report.DataSource = CreateDataSource();
            report.DataMember = "Categories";
            return report;
        }

        private object CreateDataSource()
        {
            var ds = new SqlDataSource("nwind");
            SelectQuery query = SelectQueryFluentBuilder.AddTable("Categories")
                .SelectAllColumns()
                .Build("Categories");
            ds.Queries.Add(query);
            ds.Fill();
            return ds;
        }
    }
}

File: VB/ReportAtRuntimeMvcApp/Services/CustomReportProvider.vb
Imports DevExpress.DataAccess.Sql
Imports DevExpress.XtraPrinting
Imports DevExpress.XtraReports.Services
Imports DevExpress.XtraReports.UI
Imports System.Drawing

Namespace ReportAtRuntimeMvcApp.Services

    Public Class CustomReportProvider
        Implements IReportProvider

        Public Function GetReport(ByVal id As String, ByVal context As ReportProviderContext) As XtraReport Implements IReportProvider.GetReport
            If String.IsNullOrEmpty(id) Then Return Nothing
            Select Case id
                Case "TestReport"
                    Return CreateProductsReport()
            End Select

            Return New XtraReport()
        End Function

        Private Function CreateProductsReport() As XtraReport
            Dim report As XtraReport = New XtraReport()
            Dim headerBand As ReportHeaderBand = New ReportHeaderBand() With {.HeightF = 80}
            report.Bands.Add(headerBand)
            headerBand.Controls.Add(New XRLabel() With {.Text = "Categories Report", .SizeF = New SizeF(650, 80), .TextAlignment = TextAlignment.BottomCenter, .Font = New Font("Arial", 36)})
            Dim detailBand As DetailBand = New DetailBand()
            report.Bands.Add(detailBand)
            Dim pbPicture As XRPictureBox = New XRPictureBox() With {.LocationF = New PointF(10, 10), .SizeF = New SizeF(190, 90), .Sizing = ImageSizeMode.ZoomImage}
            pbPicture.ExpressionBindings.Add(New ExpressionBinding("Image", "Picture"))
            detailBand.Controls.Add(pbPicture)
            Dim lbCategoryName As XRLabel = New XRLabel() With {.LocationF = New PointF(200, 10), .SizeF = New SizeF(440, 50), .TextAlignment = TextAlignment.BottomLeft, .Font = New Font("Arial", 24)}
            lbCategoryName.ExpressionBindings.Add(New ExpressionBinding("Text", "CategoryName"))
            detailBand.Controls.Add(lbCategoryName)
            Dim lbDescription As XRLabel = New XRLabel() With {.LocationF = New PointF(200, 60), .SizeF = New SizeF(440, 40), .TextAlignment = TextAlignment.TopLeft, .Font = New Font("Arial", 14, FontStyle.Italic)}
            lbDescription.ExpressionBindings.Add(New ExpressionBinding("Text", "Description"))
            detailBand.Controls.Add(lbDescription)
            report.DataSource = CreateDataSource()
            report.DataMember = "Categories"
            Return report
        End Function

        Private Function CreateDataSource() As Object
            Dim ds = New SqlDataSource("nwind")
            Dim query As SelectQuery = SelectQueryFluentBuilder.AddTable("Categories").SelectAllColumns().Build("Categories")
            ds.Queries.Add(query)
            ds.Fill()
            Return ds
        End Function
    End Class
End Namespace
